### PR TITLE
matrix tests

### PR DESCRIPTION
**Pull request by CpuID**
Currently supports:

- multiple values files (one to start) which will be looked up and used as input in matrix jobs
- multiple helm versions (2 to start)
- kubeconform enabled and passing
- kube-score disabled (lots of warnings atm)

### CODE REVIEW

**Comment by chris-sanders** [comment]
> this is actually one of the things I love the most about that kubeconform action, it iterates over a tests directory, runs helm template on each of the values files it finds, then feeds each into kubeconform which is kind of amazing. [source code](https://github.com/shivjm/helm-kubeconform-action/blob/main/main.go) - I'm thinking of forking that action to take a similar approach for other testing frameworks possibly :)

Yea I really like that pattern too. What would be really nice is if terratest and kubeconform could run off the same set of values files. Then github actions could run kubeconform and the user could just write terratest tests. The user wouldn't have to learn 2 tools, but we would get validation for every test case we added.

That is, if kubeconform does anything above and beyond terratest. I don't yet understand the difference in what kubeconform does vs terratest using the go client. I'm not clear if the client is already performing the same validation kubeconform is.

File: Taskfile.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    cmds:
+      - grep version charts/replicated-library/Chart.yaml | cut -d' ' -f2 > ./library_chart_version
+      - cp test-charts/test-replicated-library/Chart.yaml.tmpl test-charts/test-replicated-library/Chart.yaml
+      - 'sed -i.bak "s/^  version:.*$/  version: $(cat ./library_chart_version)/" test-charts/test-replicated-library/Chart.yaml'
+      - rm -f test-charts/test-replicated-library/Chart.yaml.bak
+      - rm -f ./library_chart_version
+  test:
+    cmds:
+      # Make sure we have a clean slate (for local use)
+      - rm -rf charts
```
Comment (chris-sanders) [inline Taskfile.yml:17]: This command seems very dangerous. Running this command from the root folder would remove all of a users work in progress that hasn't been committed.

**Review by chris-sanders — CHANGES_REQUESTED** [review]
While I don't see any particular issues with kubeconform, I'm also not particularly excited about the value this brings by itself.

The key task we need to answer is if a given values file produces the expected manifests. IE, if I specify an environment variable and a volume claim are those set, and are they set on the right pod. None of the tools shown here can answer that. A valid pod spec, without my environment variable is a test failure.

I don't think we should land any testing changes until the primary problem of testing that the values file does what we expect is addressed. The reason for that is, I am confident we can easily solve for linting in several ways. I do not want our approach to linting to get in the way of the primary testing we need to do, validating the input produces the expected output configuration.

Things like terratest likely cover linting as part of the testing removing the need for any additional linting libraries. This is also using a pattern of pre-rendering the templates and then running commands over the final output. It's not clear to me that same pattern will work when we want to verify the expected output which might change the entry point to how we run tests all together.

File: Taskfile.yml (hunk 1 above)
Comment (CpuID) [reply Taskfile.yml:17]: This refers to the directory `test-charts/<testchartname>/charts`, not the repo root `charts`.

Specifically due to this line: `dir: test-charts/{{.CHART}}`

Which technically makes it safe :)

**Pull request closed without merging**